File: frontend/src/utils/ags.ts
/**
 * AGS · Agenda de Servicios — tipos y utilidades compartidas.
 *
 * Todo el módulo habla el mismo lenguaje desde aquí: los tipos que devuelve
 * el backend, el formato de pesos y los colores de cada estado.
 */

export const AGS_COLOR = '#A21CAF'
export const AGS_DARK = '#86198F'
export const AGS_LIGHT = '#F5D0FE'

// ── Formato de dinero ──────────────────────────────────────────────────
// El peso colombiano no maneja centavos en la práctica: mostrarlos solo
// ensucia la lectura de una lista de precios.
export function fmtCOP(v?: number | null): string {
  if (v === null || v === undefined || Number.isNaN(v)) return '$0'
  return '$' + Math.round(v).toLocaleString('es-CO')
}

/** Versión compacta para tarjetas de KPI: $1,2M / $850k */
export function fmtCortoCOP(v?: number | null): string {
  if (!v) return '$0'
  const abs = Math.abs(v)
  if (abs >= 1_000_000) return '$' + (v / 1_000_000).toFixed(1).replace('.0', '') + 'M'
  if (abs >= 10_000) return '$' + Math.round(v / 1000) + 'k'
  return fmtCOP(v)
}

export function fmtMinutos(min?: number | null): string {
  if (!min) return '0 min'
  const h = Math.floor(min / 60)
  const m = min % 60
  if (h === 0) return `${m} min`
  if (m === 0) return `${h} h`
  return `${h} h ${m} min`
}

/** "24/08/2026" */
export function fmtFecha(iso?: string | null): string {
  if (!iso) return '—'
  const d = new Date(iso)
  if (Number.isNaN(d.getTime())) return '—'
  return d.toLocaleDateString('es-CO', { day: '2-digit', month: '2-digit', year: 'numeric' })
}

/** "lun 24 ago" */
export function fmtFechaCorta(iso?: string | null): string {
  if (!iso) return '—'
  const d = new Date(iso)
  if (Number.isNaN(d.getTime())) return '—'
  return d.toLocaleDateString('es-CO', { weekday: 'short', day: '2-digit', month: 'short' })
}

/** "09:00" — se lee del texto ISO y no del objeto Date para no aplicarle
 *  la zona del navegador: el backend ya entrega la hora local del negocio. */
export function fmtHora(iso?: string | null): string {
  if (!iso) return '—'
  const m = /T(\d{2}:\d{2})/.exec(iso)
  return m ? m[1] : '—'
}

export function fmtFechaHora(iso?: string | null): string {
  if (!iso) return '—'
  return `${fmtFecha(iso)} ${fmtHora(iso)}`
}

/** Fecha de hoy como YYYY-MM-DD en hora local (no UTC). */
export function hoyISO(): string {
  const d = new Date()
  const mes = String(d.getMonth() + 1).padStart(2, '0')
  const dia = String(d.getDate()).padStart(2, '0')
  return `${d.getFullYear()}-${mes}-${dia}`
}

export function sumarDiasISO(iso: string, dias: number): string {
  const d = new Date(`${iso}T12:00:00`)
  d.setDate(d.getDate() + dias)
  const mes = String(d.getMonth() + 1).padStart(2, '0')
  const dia = String(d.getDate()).padStart(2, '0')
  return `${d.getFullYear()}-${mes}-${dia}`
}

/** Primer día del mes de la fecha dada. */
export function inicioMesISO(iso = hoyISO()): string {
  return `${iso.slice(0, 7)}-01`
}

// ── Estados de la cita ─────────────────────────────────────────────────
export const ESTADOS_CITA = [
  { valor: 'AGENDADA',   label: 'Agendada',   color: '#0284C7', descripcion: 'Reservada, sin confirmar' },
  { valor: 'CONFIRMADA', label: 'Confirmada', color: '#7C3AED', descripcion: 'El cliente confirmó que viene' },
  { valor: 'EN_CURSO',   label: 'En curso',   color: '#CA8A04', descripcion: 'Se está atendiendo' },
  { valor: 'COMPLETADA', label: 'Completada', color: '#16A34A', descripcion: 'Atendida y cobrada' },
  { valor: 'CANCELADA',  label: 'Cancelada',  color: '#64748B', descripcion: 'Cancelada con aviso' },
  { valor: 'NO_ASISTIO', label: 'No asistió', color: '#DC2626', descripcion: 'El cliente no llegó' },
] as const

export function estadoCita(valor?: string | null) {
  return ESTADOS_CITA.find(e => e.valor === valor)
    ?? { valor: valor ?? '', label: valor ?? '—', color: '#64748B', descripcion: '' }
}

/** Acciones posibles según el estado actual (espeja las transiciones del backend). */
export const TRANSICIONES_CITA: Record<string, string[]> = {
  AGENDADA:   ['CONFIRMADA', 'EN_CURSO', 'CANCELADA', 'NO_ASISTIO'],
  CONFIRMADA: ['EN_CURSO', 'CANCELADA', 'NO_ASISTIO'],
  EN_CURSO:   ['COMPLETADA', 'CANCELADA'],
  COMPLETADA: [],
  CANCELADA:  [],
  NO_ASISTIO: [],
}

export const MEDIOS_PAGO = [
  { valor: 'EFECTIVO',      label: 'Efectivo' },
  { valor: 'NEQUI',         label: 'Nequi' },
  { valor: 'DAVIPLATA',     label: 'Daviplata' },
  { valor: 'TRANSFERENCIA', label: 'Transferencia' },
  { valor: 'TARJETA',       label: 'Tarjeta' },
  { valor: 'QR',            label: 'QR' },
  { valor: 'CREDITO',       label: 'Crédito (queda por cobrar)' },
] as const

export const ORIGENES_CITA = [
  { valor: 'MOSTRADOR', label: 'Mostrador' },
  { valor: 'TELEFONO',  label: 'Teléfono' },
  { valor: 'WHATSAPP',  label: 'WhatsApp' },
  { valor: 'ONLINE',    label: 'Agenda online' },
] as const

export const TIPOS_NEGOCIO = [
  { valor: 'SALON_BELLEZA', label: 'Salón de belleza' },
  { valor: 'BARBERIA',      label: 'Barbería' },
  { valor: 'SPA',           label: 'Spa / estética' },
  { valor: 'UNAS',          label: 'Salón de uñas' },
  { valor: 'PLOMERIA',      label: 'Plomería' },
  { valor: 'CONSTRUCCION',  label: 'Albañilería / remodelación' },
  { valor: 'ELECTRICIDAD',  label: 'Electricidad' },
  { valor: 'TECNICO',       label: 'Servicio técnico a domicilio' },
  { valor: 'OTRO',          label: 'Otro' },
] as const

export const DIAS_SEMANA = [
  { valor: 1, label: 'Lunes',     corto: 'Lun' },
  { valor: 2, label: 'Martes',    corto: 'Mar' },
  { valor: 3, label: 'Miércoles', corto: 'Mié' },
  { valor: 4, label: 'Jueves',    corto: 'Jue' },
  { valor: 5, label: 'Viernes',   corto: 'Vie' },
  { valor: 6, label: 'Sábado',    corto: 'Sáb' },
  { valor: 7, label: 'Domingo',   corto: 'Dom' },
] as const

// ── Tipos del backend ──────────────────────────────────────────────────

export interface AGSConfigData {
  id: number
  nombre_negocio: string
  tipo_negocio?: string | null
  nit?: string | null
  telefono?: string | null
  direccion?: string | null
  ciudad?: string | null
  hora_apertura?: string | null
  hora_cierre?: string | null
  dias_laborales?: number[] | null
  intervalo_agenda_min?: number | null
  moneda?: string | null
  iva_pct?: number | null
  comision_defecto_pct?: number | null
  permite_sobrecupo?: boolean | null
  anticipacion_minima_min?: number | null
  tolerancia_no_show_min?: number | null
  mensaje_recordatorio?: string | null
  // Reserva online (página pública /reservar/{slug})
  reserva_online_activa?: boolean | null
  slug?: string | null
  mensaje_bienvenida?: string | null
  dias_max_anticipacion?: number | null
  max_citas_pendientes_cliente?: number | null
  permite_cancelar_online?: boolean | null
  horas_min_cancelacion?: number | null
  requiere_confirmacion_online?: boolean | null
}

export interface Categoria {
  id: number
  nombre: string
  descripcion?: string | null
  color?: string | null
  orden?: number | null
  activo?: boolean | null
  total_servicios?: number
}

export interface Servicio {
  id: number
  codigo: string
  nombre: string
  categoria_id?: number | null
  categoria_nombre?: string | null
  categoria_color?: string | null
  descripcion?: string | null
  duracion_min: number
  precio: number
  costo_insumos?: number | null
  comision_pct?: number | null
  permite_domicilio?: boolean | null
  cobra_materiales?: boolean | null
  requiere_anticipo?: boolean | null
  color?: string | null
  activo?: boolean | null
  margen?: number | null
  margen_pct?: number | null
  veces_vendido?: number
}

export interface Horario {
  id?: number
  profesional_id?: number
  dia_semana: number
  hora_inicio: string
  hora_fin: string
  activo?: boolean | null
}

export interface Profesional {
  id: number
  codigo: string
  nombre: string
  documento?: string | null
  telefono?: string | null
  email?: string | null
  especialidad?: string | null
  color?: string | null
  comision_pct?: number | null
  salario_base?: number | null
  fecha_ingreso?: string | null
  acepta_domicilio?: boolean | null
  notas?: string | null
  activo?: boolean | null
  horarios?: Horario[]
  servicios_ids?: number[]
  citas_mes?: number
  ingresos_mes?: number
}

export interface Cliente {
  id: number
  codigo: string
  nombre: string
  documento?: string | null
  telefono?: string | null
  email?: string | null
  direccion?: string | null
  barrio?: string | null
  ciudad?: string | null
  fecha_nacimiento?: string | null
  como_nos_conocio?: string | null
  acepta_recordatorios?: boolean | null
  notas?: string | null
  activo?: boolean | null
  total_citas?: number
  citas_completadas?: number
  citas_no_asistio?: number
  total_gastado?: number
  ticket_promedio?: number
  ultima_visita?: string | null
  proxima_cita?: string | null
  saldo_pendiente?: number
}

export interface LineaServicio {
  id: number
  servicio_id?: number | null
  nombre_servicio: string
  cantidad: number
  precio_unitario: number
  subtotal: number
  duracion_min: number
}

export interface LineaMaterial {
  id: number
  descripcion: string
  cantidad: number
  precio_unitario: number
  subtotal: number
}

export interface Pago {
  id: number
  fecha: string
  monto: number
  medio_pago?: string | null
  tipo?: string | null
  referencia?: string | null
  notas?: string | null
}

export interface Cita {
  id: number
  codigo: string
  cliente_id: number
  cliente?: string | null
  cliente_telefono?: string | null
  profesional_id: number
  profesional?: string | null
  profesional_color?: string | null
  fecha_inicio: string
  fecha_fin: string
  duracion_min: number
  lugar?: string | null
  direccion_servicio?: string | null
  estado: string
  origen?: string | null
  subtotal: number
  descuento: number
  descuento_motivo?: string | null
  total_materiales: number
  propina: number
  total: number
  pagado: boolean
  total_pagado: number
  saldo: number
  medio_pago?: string | null
  fecha_pago?: string | null
  comision_profesional: number
  notas?: string | null
  motivo_cancelacion?: string | null
  recordatorio_enviado: boolean
  servicios_texto?: string
  servicios?: LineaServicio[]
  materiales?: LineaMaterial[]
  pagos?: Pago[]
}

export interface SlotDisponible {
  hora_inicio: string
  hora_fin: string
  inicio: string
  fin: string
}

export interface DisponibilidadProfesional {
  profesional_id: number
  profesional: string
  color?: string | null
  trabaja: boolean
  motivo_no_disponible?: string | null
  jornada: string[]
  slots: SlotDisponible[]
  minutos_disponibles: number
  minutos_ocupados: number
  ocupacion_pct: number
}

export interface Ausencia {
  id: number
  profesional_id?: number | null
  profesional_nombre?: string | null
  fecha_inicio: string
  fecha_fin: string
  motivo?: string | null
  tipo?: string | null
}

export interface PuntoIngreso {
  periodo: string
  fecha: string
  citas: number
  servicios: number
  materiales: number
  descuentos: number
  propinas: number
  total: number
  comisiones: number
  utilidad: number
}

export interface ResumenIngresos {
  desde: string
  hasta: string
  agrupar: string
  citas_completadas: number
  citas_canceladas: number
  citas_no_asistio: number
  total_servicios: number
  total_materiales: number
  total_descuentos: number
  total_propinas: number
  total_ingresos: number
  total_comisiones: number
  utilidad_bruta: number
  ticket_promedio: number
  por_cobrar: number
  tasa_no_show_pct: number
  serie: PuntoIngreso[]
}

export interface ProduccionProfesional {
  profesional_id: number
  profesional: string
  color?: string | null
  citas: number
  servicios: number
  ingresos: number
  comision: number
  propinas: number
  ticket_promedio: number
  minutos_trabajados: number
  minutos_disponibles: number
  ocupacion_pct: number
  no_show: number
}

export interface VentaServicio {
  servicio_id?: number | null
  servicio: string
  categoria?: string | null
  veces: number
  cantidad: number
  ingresos: number
  participacion_pct: number
  minutos: number
  ingreso_por_hora: number
}

export interface ClienteRanking {
  cliente_id: number
  cliente: string
  telefono?: string | null
  citas: number
  ingresos: number
  ticket_promedio: number
  ultima_visita?: string | null
  dias_sin_venir?: number | null
  saldo_pendiente: number
  no_show: number
}

export interface LineaCaja {
  medio_pago: string
  movimientos: number
  total: number
}

export interface CierreCaja {
  fecha: string
  total_recaudado: number
  movimientos: number
  efectivo: number
  digital: number
  por_medio: LineaCaja[]
  citas_atendidas: number
  citas_pendientes_pago: number
  saldo_por_cobrar: number
  comisiones_generadas: number
}

export interface CitaResumen {
  id: number
  codigo: string
  hora: string
  fecha_inicio: string
  cliente: string
  telefono?: string | null
  profesional?: string | null
  profesional_color?: string | null
  servicios: string
  estado: string
  total: number
  lugar?: string | null
}

export interface DashboardAGS {
  fecha: string
  negocio?: string | null
  citas_hoy: number
  atendidas_hoy: number
  pendientes_hoy: number
  ingresos_hoy: number
  recaudado_hoy: number
  ocupacion_hoy_pct: number
  citas_mes: number
  ingresos_mes: number
  ticket_promedio_mes: number
  comisiones_mes: number
  ingresos_mes_anterior: number
  variacion_pct?: number | null
  por_cobrar: number
  citas_por_cobrar: number
  no_show_mes: number
  tasa_no_show_pct: number
  clientes_nuevos_mes: number
  clientes_activos: number
  sin_recordatorio: number
  agenda_hoy: CitaResumen[]
  proximas: CitaResumen[]
  top_servicios: VentaServicio[]
}

export interface HistorialCliente {
  cliente_id: number
  nombre: string
  total_citas: number
  total_gastado: number
  ticket_promedio: number
  saldo_pendiente: number
  dias_desde_ultima?: number | null
  servicio_favorito?: string | null
  citas: {
    id: number
    codigo: string
    fecha_inicio: string
    estado: string
    profesional?: string | null
    servicios: string
    total: number
    total_pagado: number
    medio_pago?: string | null
  }[]
}

export interface Recordatorio {
  cita_id: number
  cliente: string
  telefono?: string | null
  mensaje: string
  enlace_whatsapp?: string | null
}

File: backend/app/infrastructure/models/ags.py
"""AGS - Agenda de Servicios.

Modulo para negocios de servicios con cita previa: salones de belleza,
barberias, spa, plomeros, albaniles, tecnicos a domicilio.

La idea central es que el negocio preconfigura su catalogo de servicios
(con precio y duracion), su equipo y los horarios de cada persona; sobre eso
la agenda calcula disponibilidad real y cada cita atendida se convierte en un
ingreso trazable por cliente, por profesional y por servicio.
"""
import enum
import sqlalchemy as sa
from sqlalchemy import UniqueConstraint
from app.core.database import Base
from app.infrastructure.models.base import TimestampMixin


# ──────────────────────────────────────────
# ENUMS
# ──────────────────────────────────────────

class EstadoCitaEnum(str, enum.Enum):
    AGENDADA   = "AGENDADA"      # reservada, sin confirmar
    CONFIRMADA = "CONFIRMADA"    # el cliente confirmo que viene
    EN_CURSO   = "EN_CURSO"      # se esta atendiendo
    COMPLETADA = "COMPLETADA"    # atendida y cobrada
    CANCELADA  = "CANCELADA"     # cancelada con aviso
    NO_ASISTIO = "NO_ASISTIO"    # el cliente no llego (no-show)


class LugarServicioEnum(str, enum.Enum):
    LOCAL     = "LOCAL"       # en el establecimiento
    DOMICILIO = "DOMICILIO"   # en la direccion del cliente


class OrigenCitaEnum(str, enum.Enum):
    MOSTRADOR = "MOSTRADOR"
    TELEFONO  = "TELEFONO"
    WHATSAPP  = "WHATSAPP"
    ONLINE    = "ONLINE"      # autoagendada por el cliente
    RECURRENTE = "RECURRENTE"  # generada desde una cita periodica


class MedioPagoEnum(str, enum.Enum):
    EFECTIVO     = "EFECTIVO"
    NEQUI        = "NEQUI"
    DAVIPLATA    = "DAVIPLATA"
    TRANSFERENCIA = "TRANSFERENCIA"
    TARJETA      = "TARJETA"
    QR           = "QR"
    CREDITO      = "CREDITO"    # queda por cobrar


class TipoPagoEnum(str, enum.Enum):
    ANTICIPO = "ANTICIPO"   # abono antes de terminar (obras, trabajos largos)
    PAGO     = "PAGO"       # pago del servicio
    PROPINA  = "PROPINA"


# ──────────────────────────────────────────
# CONFIGURACION DEL NEGOCIO
# ──────────────────────────────────────────

class AGSConfig(Base, TimestampMixin):
    """Parametros del negocio. Fila unica (id=1)."""
    __tablename__ = "ags_config"

    id = sa.Column(sa.Integer, primary_key=True, index=True)
    nombre_negocio = sa.Column(sa.String(160), nullable=False, default="Mi negocio")
    tipo_negocio   = sa.Column(sa.String(60), default="SALON_BELLEZA")
    nit            = sa.Column(sa.String(40), nullable=True)
    telefono       = sa.Column(sa.String(40), nullable=True)
    direccion      = sa.Column(sa.String(200), nullable=True)
    ciudad         = sa.Column(sa.String(80), nullable=True)

    # Ventana de atencion y granularidad de la agenda
    hora_apertura  = sa.Column(sa.String(5), default="08:00")
    hora_cierre    = sa.Column(sa.String(5), default="19:00")
    dias_laborales = sa.Column(sa.JSON, nullable=True)   # [1,2,3,4,5,6] lunes=1..domingo=7
    intervalo_agenda_min = sa.Column(sa.Integer, default=30)

    # Dinero
    moneda            = sa.Column(sa.String(8), default="COP")
    iva_pct           = sa.Column(sa.Float, default=0)
    comision_defecto_pct = sa.Column(sa.Float, default=0)

    # Politicas de agenda
    permite_sobrecupo      = sa.Column(sa.Boolean, default=False)  # permitir 2 citas a la vez
    anticipacion_minima_min = sa.Column(sa.Integer, default=0)     # no agendar con menos de X min
    tolerancia_no_show_min  = sa.Column(sa.Integer, default=15)

    # Plantilla del recordatorio de WhatsApp
    mensaje_recordatorio = sa.Column(
        sa.Text,
        default="Hola {cliente}, le recordamos su cita en {negocio} "
                "el {fecha} a las {hora} para {servicio}. Cualquier cambio nos avisa.",
    )


# ──────────────────────────────────────────
# CATALOGO DE SERVICIOS
# ──────────────────────────────────────────

class AGSCategoriaServicio(Base, TimestampMixin):
    __tablename__ = "ags_categoria_servicio"
    __table_args__ = (UniqueConstraint("nombre", name="uq_ags_categoria_nombre"),)

    id     = sa.Column(sa.Integer, primary_key=True, index=True)
    nombre = sa.Column(sa.String(120), nullable=False)
    descripcion = sa.Column(sa.Text, nullable=True)
    color  = sa.Column(sa.String(9), default="#A21CAF")
    orden  = sa.Column(sa.Integer, default=0)
    activo = sa.Column(sa.Boolean, default=True, nullable=False)


class AGSServicio(Base, TimestampMixin):
    """Servicio con precio y duracion preconfigurados.

    La duracion es lo que permite que la agenda calcule la hora de fin y
    detecte cruces; el precio es lo que alimenta los ingresos.
    """
    __tablename__ = "ags_servicio"
    __table_args__ = (UniqueConstraint("codigo", name="uq_ags_servicio_codigo"),)

    id     = sa.Column(sa.Integer, primary_key=True, index=True)
    codigo = sa.Column(sa.String(40), nullable=False, index=True)
    nombre = sa.Column(sa.String(160), nullable=False)
    categoria_id = sa.Column(sa.Integer, sa.ForeignKey("ags_categoria_servicio.id", ondelete="SET NULL"), nullable=True)
    descripcion  = sa.Column(sa.Text, nullable=True)

    duracion_min = sa.Column(sa.Integer, nullable=False, default=30)
    precio       = sa.Column(sa.Float, nullable=False, default=0)
    costo_insumos = sa.Column(sa.Float, default=0)      # para calcular margen real
    comision_pct  = sa.Column(sa.Float, nullable=True)  # override de la comision del profesional

    # Los oficios a domicilio (plomeria, albanileria) necesitan direccion y
    # suelen cobrar materiales aparte de la mano de obra.
    permite_domicilio = sa.Column(sa.Boolean, default=False)
    cobra_materiales  = sa.Column(sa.Boolean, default=False)
    requiere_anticipo = sa.Column(sa.Boolean, default=False)

    color  = sa.Column(sa.String(9), nullable=True)
    activo = sa.Column(sa.Boolean, default=True, nullable=False)


# ──────────────────────────────────────────
# EQUIPO
# ──────────────────────────────────────────

class AGSProfesional(Base, TimestampMixin):
    """Quien presta el servicio: estilista, barbero, plomero, oficial."""
    __tablename__ = "ags_profesional"
    __table_args__ = (UniqueConstraint("codigo", name="uq_ags_profesional_codigo"),)

    id     = sa.Column(sa.Integer, primary_key=True, index=True)
    codigo = sa.Column(sa.String(40), nullable=False, index=True)
    nombre = sa.Column(sa.String(160), nullable=False)
    documento = sa.Column(sa.String(40), nullable=True)
    telefono  = sa.Column(sa.String(40), nullable=True)
    email     = sa.Column(sa.String(160), nullable=True)
    especialidad = sa.Column(sa.String(120), nullable=True)

    color = sa.Column(sa.String(9), default="#A21CAF")   # para distinguirlo en la agenda
    comision_pct = sa.Column(sa.Float, default=0)
    salario_base = sa.Column(sa.Float, default=0)
    fecha_ingreso = sa.Column(sa.Date, nullable=True)

    acepta_domicilio = sa.Column(sa.Boolean, default=False)
    notas  = sa.Column(sa.Text, nullable=True)
    activo = sa.Column(sa.Boolean, default=True, nullable=False)


class AGSProfesionalServicio(Base, TimestampMixin):
    """Que servicios sabe prestar cada profesional."""
    __tablename__ = "ags_profesional_servicio"
    __table_args__ = (
        UniqueConstraint("profesional_id", "servicio_id", name="uq_ags_prof_servicio"),
    )

    id = sa.Column(sa.Integer, primary_key=True, index=True)
    profesional_id = sa.Column(sa.Integer, sa.ForeignKey("ags_profesional.id", ondelete="CASCADE"), nullable=False, index=True)
    servicio_id    = sa.Column(sa.Integer, sa.ForeignKey("ags_servicio.id", ondelete="CASCADE"), nullable=False, index=True)


class AGSHorarioProfesional(Base, TimestampMixin):
    """Jornada habitual por dia de la semana. Base del calculo de disponibilidad.

    Se permiten varias franjas por dia para modelar la hora de almuerzo
    (ej. 08:00-12:00 y 14:00-18:00).
    """
    __tablename__ = "ags_horario_profesional"
    __table_args__ = (
        UniqueConstraint("profesional_id", "dia_semana", "hora_inicio", name="uq_ags_horario_prof_dia"),
    )

    id = sa.Column(sa.Integer, primary_key=True, index=True)
    profesional_id = sa.Column(sa.Integer, sa.ForeignKey("ags_profesional.id", ondelete="CASCADE"), nullable=False, index=True)
    dia_semana  = sa.Column(sa.Integer, nullable=False)   # 1=lunes ... 7=domingo
    hora_inicio = sa.Column(sa.String(5), nullable=False)  # "08:00"
    hora_fin    = sa.Column(sa.String(5), nullable=False)  # "12:00"
    activo = sa.Column(sa.Boolean, default=True, nullable=False)


class AGSAusencia(Base, TimestampMixin):
    """Bloqueo de agenda: vacaciones, incapacidad, permiso, dia festivo."""
    __tablename__ = "ags_ausencia"

    id = sa.Column(sa.Integer, primary_key=True, index=True)
    profesional_id = sa.Column(sa.Integer, sa.ForeignKey("ags_profesional.id", ondelete="CASCADE"), nullable=True, index=True)
    # profesional_id nulo = cierra el negocio completo (festivo, mantenimiento)
    fecha_inicio = sa.Column(sa.DateTime(timezone=True), nullable=False, index=True)
    fecha_fin    = sa.Column(sa.DateTime(timezone=True), nullable=False)
    motivo = sa.Column(sa.String(160), nullable=True)
    tipo   = sa.Column(sa.String(40), default="PERMISO")


# ──────────────────────────────────────────
# CLIENTES
# ──────────────────────────────────────────

class AGSCliente(Base, TimestampMixin):
    __tablename__ = "ags_cliente"
    __table_args__ = (UniqueConstraint("codigo", name="uq_ags_cliente_codigo"),)

    id     = sa.Column(sa.Integer, primary_key=True, index=True)
    codigo = sa.Column(sa.String(40), nullable=False, index=True)
    nombre = sa.Column(sa.String(160), nullable=False, index=True)
    documento = sa.Column(sa.String(40), nullable=True)
    telefono  = sa.Column(sa.String(40), nullable=True, index=True)
    email     = sa.Column(sa.String(160), nullable=True)
    direccion = sa.Column(sa.String(200), nullable=True)
    barrio    = sa.Column(sa.String(120), nullable=True)
    ciudad    = sa.Column(sa.String(80), nullable=True)
    fecha_nacimiento = sa.Column(sa.Date, nullable=True)

    como_nos_conocio = sa.Column(sa.String(80), nullable=True)
    acepta_recordatorios = sa.Column(sa.Boolean, default=True)
    notas  = sa.Column(sa.Text, nullable=True)   # alergias, preferencias, tono de tinte
    activo = sa.Column(sa.Boolean, default=True, nullable=False)


# ──────────────────────────────────────────
# CITAS
# ──────────────────────────────────────────

class AGSCita(Base, TimestampMixin):
    __tablename__ = "ags_cita"
    __table_args__ = (UniqueConstraint("codigo", name="uq_ags_cita_codigo"),)

    id     = sa.Column(sa.Integer, primary_key=True, index=True)
    codigo = sa.Column(sa.String(40), nullable=False, index=True)

    cliente_id     = sa.Column(sa.Integer, sa.ForeignKey("ags_cliente.id", ondelete="RESTRICT"), nullable=False, index=True)
    profesional_id = sa.Column(sa.Integer, sa.ForeignKey("ags_profesional.id", ondelete="RESTRICT"), nullable=False, index=True)

    fecha_inicio = sa.Column(sa.DateTime(timezone=True), nullable=False, index=True)
    fecha_fin    = sa.Column(sa.DateTime(timezone=True), nullable=False)
    duracion_min = sa.Column(sa.Integer, nullable=False, default=30)

    lugar = sa.Column(sa.String(20), default=LugarServicioEnum.LOCAL.value)
    direccion_servicio = sa.Column(sa.String(200), nullable=True)

    estado = sa.Column(sa.String(20), default=EstadoCitaEnum.AGENDADA.value, nullable=False, index=True)
    origen = sa.Column(sa.String(20), default=OrigenCitaEnum.MOSTRADOR.value)

    # Dinero: subtotal de servicios + materiales - descuento + propina = total
    subtotal         = sa.Column(sa.Float, default=0)
    descuento        = sa.Column(sa.Float, default=0)
    descuento_motivo = sa.Column(sa.String(160), nullable=True)
    total_materiales = sa.Column(sa.Float, default=0)
    propina          = sa.Column(sa.Float, default=0)
    total            = sa.Column(sa.Float, default=0)

    pagado     = sa.Column(sa.Boolean, default=False, index=True)
    total_pagado = sa.Column(sa.Float, default=0)
    medio_pago = sa.Column(sa.String(20), nullable=True)
    fecha_pago = sa.Column(sa.DateTime(timezone=True), nullable=True)

    comision_profesional = sa.Column(sa.Float, default=0)

    notas = sa.Column(sa.Text, nullable=True)
    motivo_cancelacion = sa.Column(sa.String(200), nullable=True)
    recordatorio_enviado = sa.Column(sa.Boolean, default=False)
    creado_por = sa.Column(sa.String(120), nullable=True)

    # Fechas reales de atencion (para medir puntualidad y duracion efectiva)
    hora_llegada = sa.Column(sa.DateTime(timezone=True), nullable=True)
    hora_inicio_real = sa.Column(sa.DateTime(timezone=True), nullable=True)
    hora_fin_real    = sa.Column(sa.DateTime(timezone=True), nullable=True)


class AGSCitaServicio(Base, TimestampMixin):
    """Servicios incluidos en la cita.

    Guarda copia del nombre y del precio del momento: si manana suben la
    tarifa, los ingresos historicos no cambian.
    """
    __tablename__ = "ags_cita_servicio"

    id = sa.Column(sa.Integer, primary_key=True, index=True)
    cita_id     = sa.Column(sa.Integer, sa.ForeignKey("ags_cita.id", ondelete="CASCADE"), nullable=False, index=True)
    servicio_id = sa.Column(sa.Integer, sa.ForeignKey("ags_servicio.id", ondelete="SET NULL"), nullable=True)

    nombre_servicio = sa.Column(sa.String(160), nullable=False)
    cantidad        = sa.Column(sa.Float, default=1)
    precio_unitario = sa.Column(sa.Float, default=0)
    subtotal        = sa.Column(sa.Float, default=0)
    duracion_min    = sa.Column(sa.Integer, default=0)
    comision_pct    = sa.Column(sa.Float, nullable=True)


class AGSCitaMaterial(Base, TimestampMixin):
    """Materiales cobrados al cliente (tuberia, cemento, tinte, keratina)."""
    __tablename__ = "ags_cita_material"

    id = sa.Column(sa.Integer, primary_key=True, index=True)
    cita_id = sa.Column(sa.Integer, sa.ForeignKey("ags_cita.id", ondelete="CASCADE"), nullable=False, index=True)
    descripcion     = sa.Column(sa.String(200), nullable=False)
    cantidad        = sa.Column(sa.Float, default=1)
    precio_unitario = sa.Column(sa.Float, default=0)
    subtotal        = sa.Column(sa.Float, default=0)


class AGSPagoCita(Base, TimestampMixin):
    """Cada movimiento de dinero de la cita: anticipos y pago final.

    Tenerlos por separado permite trabajos largos que se pagan por partes
    (una obra de albanileria) y cuadrar la caja por medio de pago.
    """
    __tablename__ = "ags_pago_cita"

    id = sa.Column(sa.Integer, primary_key=True, index=True)
    cita_id = sa.Column(sa.Integer, sa.ForeignKey("ags_cita.id", ondelete="CASCADE"), nullable=False, index=True)
    fecha   = sa.Column(sa.DateTime(timezone=True), nullable=False, index=True)
    monto   = sa.Column(sa.Float, nullable=False, default=0)
    medio_pago = sa.Column(sa.String(20), default=MedioPagoEnum.EFECTIVO.value)
    tipo       = sa.Column(sa.String(20), default=TipoPagoEnum.PAGO.value)
    referencia = sa.Column(sa.String(80), nullable=True)   # numero de aprobacion / comprobante
    notas      = sa.Column(sa.Text, nullable=True)
    registrado_por = sa.Column(sa.String(120), nullable=True)
